src/signature.rs

```diff
@@ -164,136 +171,301 @@ impl<const N: usize> Signature<N> {
+                        }
+
+                        while i < N {
+                            if *scan.add(i) & *mask != *needle {
+                                cursor +=
+                                    skip_offsets[*scan.add(N.saturating_sub(1)) as usize] as usize;
+                                continue 'outer;
+                            } else {
+                                mask = mask.add(1);
+                                needle = needle.add(1);
+                                i += 1;
+                            }
+                        }
+
+                        let current_cursor = cursor;
+                        cursor = cursor + 1;
+                        return Some(current_cursor);
                     }
-                    let offset = skip_offsets[scan[end] as usize];
-                    current += offset as usize;
                 }
-                None
             }
-        }
+        })
+        .fuse()
     }
+}
+
+/// Trait that provides scanning methods for the `Signature` type.
+pub trait SignatureScanner {
+    /// Scans a process's memory in the given range for the first occurrence of the signature.
+    ///
+    /// # Arguments
+    ///
+    /// * `process` - A reference to the `Process` in which the scan occurs.
+    /// * `addr` - The starting address of the memory range.
+    /// * `len` - The length of the memory range to scan.
+    ///
+    /// Returns `Some(Address)` of the first match if found, otherwise `None`.
+    fn scan<'a>(&'a self, process: &Process, addr: impl Into<Address>, len: u64)
+        -> Option<Address>;
 
-    /// Scans a process for the signature. This will scan the address range of
-    /// the process given. If the signature is found, the address of the start
-    /// of the signature is returned.
-    pub fn scan_process_range(
+    /// Returns an iterator over all occurrences of the signature in the process's memory range.
+    ///
+    /// # Arguments
+    ///
+    /// * `process` - A reference to the `Process` in which the scan occurs.
+    /// * `addr` - The starting address of the memory range.
+    /// * `len` - The length of the memory range to scan.
+    ///
+    /// Returns an iterator that yields each matching address.
+    fn scan_process_range<'a>(
+        &'a self,
+        process: &'a Process,
+        addr: impl Into<Address>,
+        len: u64,
+    ) -> impl Iterator<Item = Address> + 'a;
+
+    /// Asynchronously awaits scanning a process for the signature until a match
+    /// is found.
+    ///
+    /// # Arguments
+    ///
+    /// * `process` - A reference to the `Process` in which the scan occurs.
+    /// * `addr` - The starting address of the memory range.
+    /// * `len` - The length of the memory range to scan.
+    #[allow(async_fn_in_trait)]
+    async fn wait_scan_process_range(
         &self,
         process: &Process,
-        (addr, len): (impl Into<Address>, u64),
+        addr: impl Into<Address>,
+        len: u64,
```

Comment:
It would be better to keep the `addr` and `len` grouped into a single `range` tuple, not separate them and force every caller to separate them.

Reply:
I've reintroduced the range tuple as suggested. Thinking about it, it is also better as it doesn't break already written code.